packages/next/src/shared/lib/router/adapters.tsx

```diff
@@ -39,42 +41,35 @@ export function adaptForAppRouterInstance(
   }
 }
 
-/**
- * transforms the ParsedUrlQuery into a URLSearchParams.
- *
- * @param query the query to transform
- * @returns URLSearchParams
- */
-function transformQuery(query: ParsedUrlQuery): URLSearchParams {
-  const params = new URLSearchParams()
-
-  for (const [name, value] of Object.entries(query)) {
-    if (Array.isArray(value)) {
-      for (const val of value) {
-        params.append(name, val)
-      }
-    } else if (typeof value !== 'undefined') {
-      params.append(name, value)
-    }
-  }
-
-  return params
-}
-
 /**
  * adaptForSearchParams transforms the ParsedURLQuery into URLSearchParams.
  *
  * @param router the router that contains the query.
  * @returns the search params in the URLSearchParams format
  */
 export function adaptForSearchParams(
-  router: Pick<NextRouter, 'isReady' | 'query'>
+  router: Pick<NextRouter, 'isReady' | 'query' | 'asPath'>
 ): URLSearchParams {
   if (!router.isReady || !router.query) {
     return new URLSearchParams()
   }
 
-  return transformQuery(router.query)
+  return asPathToSearchParams(router.asPath)
```

Comment:
Why not router.query? It seems this would break with rewrites.

Reply:
search params will contain path params (like /[slug]) in router.query if we're using route.query